packages/workflow/src/workflow.ts

```diff
@@ -1590,6 +1592,49 @@ export function allHandlersFinished(): boolean {
   return activator.inProgressSignals.size === 0 && activator.inProgressUpdates.size === 0;
 }
 
+/**
+ * Can be used to alter or define workflow functions with certain options specified at definition
+ * time. In order to ensure that workflows are loaded properly by their name, you typically will not
+ * need to use the return value of this function.
+ *
+ * @example
+ * For example:
+ * ```ts
+ * defineWorkflowWithOptions({ versioningBehavior: 'pinned' }, myWorkflow);
+ * export async function myWorkflow(): Promise<string> {
+ *   // Workflow code here
+ *   return "hi";
+ * }
+ * ```
+ *
+ * @example
+ * To annotate a default or dynamic workflow:
+ * ```ts
+ * export default defineWorkflowWithOptions({ versioningBehavior: 'pinned' }, myDefaultWorkflow);
+ * async function myDefaultWorkflow(): Promise<string> {
+ *   // Workflow code here
+ *   return "hi";
+ * }
+ * ```
+ *
+ * @param options Options for the workflow defintion, or a function that returns options. If a
+ * function is provided, it will be called once just before the workflow function is called for the
+ * first time.
+ * @param fn The workflow function.
+ * @returns The same passed in workflow function, with the specified options applied. You can export
+ * this function to make it available as a workflow function.
+ */
+export function defineWorkflowWithOptions<A extends any[], RT>(
+  options: WorkflowDefinitionOptionsOrGetter,
+  fn: (...args: A) => Promise<RT>
+): WorkflowFunctionWithOptions<A, RT> {
+  const wrappedFn = Object.assign(fn, {
+    options,
+    __temporal_is_workflow_function_with_options: true as const,
```

Comment:
Why do you need this marker? Can't you simply look for `options`? Or if this is to prevent conflicting these `options` with something else that would happen to be called the same, you can instead change the name of the `options` property to something more unique.

Reply:
Ok, I suppose I can. It just feels slightly safer to have something like this but I suppose in this case it's not like there will be other collisions

Reply:
If we're really worried about collision, then using a symbol would be preferable. But I honestly don't think that's justified in this specific case. In fact, I know that many users do similar things for various reasons, so it is very likely that one may want to merge that functionality with their own wrapping logic.